tests/dag_processing/test_processor.py

```diff
@@ -1008,6 +1010,14 @@ def test_nullbyte_exception_handling_when_preimporting_airflow(self, mock_contex
         )
         processor.start()
 
+    def test_counter_for_last_num_of_db_queries(self):
+        dag_filepath = TEST_DAG_FOLDER / "test_multiple_dags.py"
+
+        with create_session() as session:
+            _, _, last_num_of_db_queries = self._process_file(dag_filepath, TEST_DAG_FOLDER, session)
+
+        assert last_num_of_db_queries == 94
```

Comment:
This unit test may be tedious to maintain.
IMHO it would be good to have a dedicated DAG for this test, which we know exactly how many queries it should do to the database. That would be also more explicit.

AFAIU this DAG test_multiple_dags.py is used by other unit tests, and it means it can be modified. If we have separate DAG for this test we will isolate this unit test from others and have better maintenance.

Reply:
FYI. We had similar issue in the past with our "do not cause unreasonable number of queries" and yes, it's been tedious to maintain as you say - also because underlying queries/code might change, so my recommmendation (on top of dedicated DAG) is to add margin to account for small changes.

See here where we have dedicated decorator for it https://github.com/apache/airflow/blob/main/tests/test_utils/asserts.py#L122

Reply:
@kosteev @potiuk I have updated the test and created a new file for DAGs. But during updating I have noticed that we count queries in a different way in my implementation and implementation which the community is using for a test. In realization for a test we apply the `after_cursor_execute` event to the `engine` object. In my realization I apply the `do_orm_execute` event to the `session`. As a result we have different count numbers.
For example these are numbers for a file with 5 DAGs with `EmptyOperator`:
- session: 26 queries for parsing a file with one DAG, 17 queries more for each new DAG. 94 for file with 5 DAGs; 
- engine: 32 queries for parsing a file with one DAG, 22 queries more for each new DAG; 120 for file with 5 DAGs; 
In that case what do you think is better to use as a counting implementation?

Reply:
I actually had a question about this in other comment, and you were saying that do_orm_execute event on the session object is better/more correct way, right?
https://github.com/apache/airflow/pull/40323#discussion_r1646359819

Reply:
My vote would be to use whatever is correct or let's say more correct:)

If I am an Airflow DAG developer, I am interested in what is exact number of queries were made during parsing of my DAG file (and I do not really care how it is calculated).

In the other comment you convinced me that Session object is better to use than Engine object in our case, now you start to question this. I am a bit confused, what's changed:)

My guess would be that counting queries on Session object should be more precise, but I would do some investigation/exploration on what is the difference before commit. So, personally I do not really know for sure, but I think counting queries on Session that is dedicated for file DAG parsing has to be correct way.

Reply:
I have adapted the `assert` function for unit tests to work with Session.